### PR TITLE
EKF API and test

### PR DESCRIPTION
**Pull request by hungwn2**

### CODE REVIEW

File: libraries/ms-common/inc/ekf.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#pragma once
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:1]: Where is the file header? Reference other header files and add it please

File: libraries/ms-common/inc/ekf.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include <stdio.h>
+#include <math.h>
+#define STATE_SIZE 9
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:4]: Please leave a comment explaining what this is

/** @brief   [DESC] */

File: libraries/ms-common/inc/ekf.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#define MEASUREMENT_SIZE 6 //(x, y, z), (ax, ay, az), (a, b, y)
+#define DT 1
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:6]: What is DT? Leave a comment like above

/** @brief [DESC] */

File: libraries/ms-common/inc/ekf.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#pragma once
+#include <stdio.h>
+#include <math.h>
+#define STATE_SIZE 9
+#define MEASUREMENT_SIZE 6 //(x, y, z), (ax, ay, az), (a, b, y)
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:5]: Please reformat the comment. Not sure what the comment means either, don't really see the relation?

File: libraries/ms-common/inc/ekf.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+#define DT 1
+
+
+
+
+//state matrix
+double x[STATE_SIZE]={};
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:12]: Why are these variables in the header file? Can we move them to the C file

File: libraries/ms-common/inc/ekf.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    {1, 0, 0, DT, 0, 0,  0,  0,  0},
+    {0, 1, 0, 0, DT, 0,  0,  0,  0},
+    {0, 0, 1, 0, 0, DT,  0,  0,  0},
+    {0, 0, 0, 1, 0, 0,   0,  0,  0},
+    {0, 0, 0, 0, 1, 0,   0,  0,  0},
+    {0, 0, 0, 0, 0, 1,   0,  0,  0},
+    {0, 0, 0, 0, 0, 0,   1, DT, 0},
+    {0, 0, 0, 0, 0, 0,   0, 1, DT},
+    {0, 0, 0, 0, 0, 0,   0, 0, 1}
+};//iniitialzie with time derivatives
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:43]: Leave a better comment on this. Please put it in the C file too

File: libraries/ms-common/inc/ekf.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+void multiply_matrices(double firstMatrix[][STATE_SIZE], double secondMatrix[][STATE_SIZE], double result[][STATE_SIZE]);
+void transponse_matrix(double matrix[][STATE_SIZE], double transpose[][STATE_SIZE]);
+
+
+void noise_covariance();
+void predict_state();
+void predict_prev_state();//pk-1|k-1
+void predict_covariance();//pk|k-1
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:56]: Add proper comments that adhere to the other files

/**
 * @brief  [FUNC DESC]] etc.
 */

File: libraries/ms-common/src/ekf.c [hunk 8]
```diff
@@ -0,0 +1,151 @@
+
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:1]: Where is the C file header? Refer to other c files

File: libraries/ms-common/src/ekf.c [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+#include <stdlib.h>
+#include "ekf.h"
+#include "log.h"
+
+
+
+
+void matrix_mult(double A[][STATE_SIZE], double B[][STATE_SIZE], double result[][STATE_SIZE]) {
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:10]: Make this a StatusCode and add checks to see if matrix A,B, or result are NULL. If they are null, return STATUS_CODE_INVALID_ARGS

File: libraries/ms-common/src/ekf.c [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            result[i][j] = 0;
+            for (int k = 0; k < STATE_SIZE; k++) {
+                result[i][j] += A[i][k] * B[k][j];
+            }
+        }
+    }
+}
+
+
+void matrix_transpose(double matrix[][STATE_SIZE], double transpose[][STATE_SIZE]){
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:22]: Make this a StatusCode and add checks to see if matrix or transpose are NULL. If they are null, return STATUS_CODE_INVALID_ARGS

File: libraries/ms-common/src/ekf.c [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    for (int i=0; i<STATE_SIZE; i++){
+        for (int j=0; j<STATE_SIZE; j++){
+            transpose[i][j]=matrix[j][i];
+        }
+    }
+}
+//use kinematics eqns to predict state
+void predict_state(){
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:30]: What eqns? Please explain the math in your comments

File: libraries/ms-common/src/ekf.c [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        for (int j = 0; j < STATE_SIZE; j++) {
+            P[i][j] = P_temp[i][j];
+        }
+    }
+}
+
+
+
+
+int main(){
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:140]: Why is there a main function here? This is an API/Library there should be no main functions

File: libraries/ms-common/test/test_ekf.c [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+    for (int i = 0; i < STATE_SIZE; i++) {
+        TEST_ASSERT_EQUAL(1.0 + Q[i][i], P[i][i]);
+    }
+
+
+}
+
+
+/* Valid FSM State Transition */
```
Comment (Akashem06) [inline libraries/ms-common/test/test_ekf.c:138]: Wrong comment

File: libraries/ms-common/inc/ekf.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+double P[STATE_SIZE][STATE_SIZE]={};
+//input controls (ax, ay, az)
+double U[MEASUREMENT_SIZE];
+
+
+//process noise and noise state covariance matrix
+double Q[STATE_SIZE][STATE_SIZE]={}; //9x9 matrix w/covariance in position, velocity, and orientation angle noise
+
+
+double R[MEASUREMENT_SIZE][MEASUREMENT_SIZE]=
```
Comment (Akashem06) [inline libraries/ms-common/inc/ekf.h:23]: You have a lot of matrices. Can you put them all into a storage struct? Having them stored together in a storage will make it much cleaner

File: libraries/ms-common/src/ekf.c [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+
+void update_state() {
+    double H[MEASUREMENT_SIZE][STATE_SIZE] = {0}; // map state vector to measurements
+    double K[STATE_SIZE][MEASUREMENT_SIZE] = {0}; //kalman gain
+    double S[MEASUREMENT_SIZE][MEASUREMENT_SIZE] = {0};//uncertainty in residual
+    double S_inv[MEASUREMENT_SIZE][MEASUREMENT_SIZE] = {0};
+    double H_T[STATE_SIZE][MEASUREMENT_SIZE] = {0};
+    double P_H_T[STATE_SIZE][MEASUREMENT_SIZE] = {0};
+    double y[MEASUREMENT_SIZE] = {0};
```
Comment (Akashem06) [inline libraries/ms-common/src/ekf.c:67]: Are all of these required? Or can some of them be reused. Your API is for embedded devices, we don't want to cause a stack overflow by allocating too much memory from 1 function call

File: libraries/ms-common/test/test_ekf.c [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+TEST_IN_TASK
+
+
+void test_matrix_transpose(void){
+    double A[STATE_SIZE][STATE_SIZE] = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
+    double expected[STATE_SIZE][STATE_SIZE];
+    double result{STATE_SIZE}[STATE_SIZE];
+    transponse_matrix(A, expected);
+    TEST_ASSERT_TRUE(compare_matrices(result, expected));
+}
```
Comment (Akashem06) [inline libraries/ms-common/test/test_ekf.c:68]: This test doesn't make sense. Youre transposing A, putting the result into 'expected', but not doing anything with 'result'? I don't think this is actually testing the code

File: libraries/ms-common/test/test_ekf.c [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+
+
+    predict_state();
+
+
+    // Expected results after one step
+    TEST_ASSERT_FLOAT_WITHIN(EPSILON, 1 + 0.5 * 1 * DT * DT, x[0]);
+    TEST_ASSERT_FLOAT_WITHIN(EPSILON, 1 + 0.5 * 2 * DT * DT, x[1]);
+    TEST_ASSERT_FLOAT_WITHIN(EPSILON, 1 + 0.5 * 3 * DT * DT, x[2]);
+}
```
Comment (Akashem06) [inline libraries/ms-common/test/test_ekf.c:111]: You copy pasted the exact same test twice

File: libraries/ms-common/test/test_ekf.c [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        P[i][i] = 1.0;  // Identity matrix for covariance
+    }
+
+
+    update_state(z, H);
+    for (int i = 0; i < MEASUREMENT_SIZE; i++) {
+        TEST_ASSERT_FLOAT_WITHIN(EPSILON, expected_x[i], x[i]);
+    }
+
+}
```
Comment (Akashem06) [inline libraries/ms-common/test/test_ekf.c:165]: You need more test cases. You need to cover the edge cases and use non-standard input values. You should also be isolating diferent situations, for example a measurement with no y or z position. and only x position. Or a test with only velocity and 0,0,0 for position. etc.

**Review by Akashem06 — CHANGES_REQUESTED** [review]
